### PR TITLE
comment changes

### PR DESCRIPTION
**Pull request by cabreraem**
I tried to fix most of what I remembered the comments to be. Enum, however, was giving me trouble and was giving me import errors so as of now I have global variables set until I can make the proper enum classes.

### CODE REVIEW

File: file_adapter.py [hunk 1]
```diff
@@ -1,28 +1,52 @@
...
-    json_data = open(tasks_file).read()
-    self.tasks = json.loads(json_data)
+    with open(releases_file) as f:
+      json_releases = f.read()
+
+    releases_dict = json.loads(json_releases)
+    self.releases = {}
+    for key in releases_dict:
+      release = Release(key)
+      for k in releases_dict[key]:
```
Comment (sebastienvas) [inline file_adapter.py]:
```python
for k, v in releases_dict[key].iteritems():
  setattr(release, k, v)
```

Comment (sebastienvas) [reply file_adapter.py]: you can use that for all iteration over a dict

File: file_adapter.py [hunk 2]
```diff
@@ -1,28 +1,52 @@
...
   def __init__(self, releases_file, tasks_file):
-    json_data = open(releases_file).read()
-    self.releases = json.loads(json_data)
-    json_data = open(tasks_file).read()
-    self.tasks = json.loads(json_data)
+    with open(releases_file) as f:
+      json_releases = f.read()
+
+    releases_dict = json.loads(json_releases)
+    self.releases = {}
```
Comment (sebastienvas) [inline file_adapter.py]: Maybe move the json to release code in the Release class. ie create a from_json method

File: app/scripts/controllers.js [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
 /* Controllers */
 
+var site = 'http://localhost:8080';
```
Comment (hklai) [inline app/scripts/controllers.js]: why is this hardcoded to localhost?

File: release_data_abc.py [hunk 4]
```diff
@@ -2,6 +2,57 @@
...
+    self.name = release_name
+    self.tasks = []
+    self.links = []
+    self.started = 0
+    self.labels = []
+    self.state = UNUSED_STATUS
+    self.last_modified = 0
+    self.last_active_task = ""
+
+  def from_dict(self, dict):
```
Comment (hklai) [inline release_data_abc.py]: This is just one level of conversion. What if a value is also a dict?

File: tests/file_adapter_tests.py [hunk 5]
```diff
@@ -13,15 +13,15 @@ def test_get_releases(self):
     releases = self.adapter.get_releases()
     length = 0
     for _, value in releases.iteritems():
-      self.assertEqual(len(value), 10)
+      print value
```
Comment (hklai) [inline tests/file_adapter_tests.py]: why print instead of assert?

File: app/scripts/controllers.js (hunk 3 above)
Comment (cabreraem) [reply app/scripts/controllers.js]: right now that's the website URL but we can change it when we push to the appengine site.

File: tests/file_adapter_tests.py (hunk 5 above)
Comment (cabreraem) [reply tests/file_adapter_tests.py]: fixed this

File: release_data_abc.py (hunk 4 above)
Comment (cabreraem) [reply release_data_abc.py]: based off of our current data model releases shouldn't have any values that are dicts, only single values and arrays

File: adapter_abc.py [hunk 6]
```diff
@@ -13,11 +13,21 @@ def get_releases(self):
     """Retrieve all release information."""
     pass
 
+  @abc.abstractmethod
+  def get_release(self, release_name):
+    """Retrieve individual release information."""
```
Comment (sebastienvas) [inline adapter_abc.py]: also describe the return object(s).

File: adapter_abc.py [hunk 7]
```diff
@@ -13,11 +13,21 @@ def get_releases(self):
...
+    pass
+
   @abc.abstractmethod
```
Comment (sebastienvas) [inline adapter_abc.py]: same comment here.

File: app/scripts/controllers.js (hunk 3 above)
Comment (sebastienvas) [reply app/scripts/controllers.js]: I am thinking there must be a deterministic way to figure that out. Since the js is downloaded from the backend.

Comment (cabreraem) [reply app/scripts/controllers.js]: you're right, I fixed this

File: adapter_abc.py (hunk 6 above)
Comment (cabreraem) [reply adapter_abc.py]: Done.

File: adapter_abc.py (hunk 7 above)
Comment (cabreraem) [reply adapter_abc.py]: Done.

File: task.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+  def __init__(self, data=None):
+    """Initializes a task.
+
+    Args:
+      data: dictionary with all the task data (optional).
+    """
+    if data is None:
+      self._task = {}
+    else:
+      self._task = data
```
Comment (hklai) [inline task.py]: self._task = data or {}

means the same for these 4 lines

File: task.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+  @property
+  def task_name(self):
+    return self._task['task_name']
+
+  @task_name.setter
+  def task_name(self, value):
+    """Sets name as string value."""
+    if isinstance(value, basestring):
```
Comment (hklai) [inline task.py]: Create a helper (parameterize the key and type) since this is used often...

File: task.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+  @error.setter
+  def error(self, value):
+    """Sets error as string."""
+    if isinstance(value, basestring):
+      self._task['error'] = value
+    else:
+      raise ValueError('Invalid input for task error: not a string')
+
+  def to_dict(self):
+    return self._task
```
Comment (hklai) [inline task.py]: Perhaps make a copy instead for better abstraction, so the caller cannot mess up internal states of this task)

File: task.py [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+      self._task['task_name'] = value
+    else:
+      raise ValueError('Invalid input for task name: not a string')
+
+  @property
+  def dependent_on(self):
+    return self._task['dependent_on']
+
+  @dependent_on.setter
+  def dependent_on(self, value):
```
Comment (hklai) [inline task.py]: perhaps better to have a add_depend_on() that takes a single object instead of an implicit list.

File: task.py (hunk 10 above)
Comment (cabreraem) [reply task.py]: so set a variable like task = self._task and then return task?

Comment (hklai) [reply task.py]: No

return self._task.copy() for a shallow copy, or using copy.deepcopy() which probably is an overkill...

File: release.py [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+  def links(self):
+    return self._release['links']
+
+  @links.setter
+  def links(self, value):
+    """Sets links as list of string urls."""
+    self._validate_array(basestring, value, 'links')
+
+  @property
+  def started(self):
```
Comment (sebastienvas) [inline release.py:86]: do we want to return a datetime here ?

File: release.py [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+  def state(self):
+    return self._release['state']
+
+  @state.setter
+  def state(self, value):
+    """Sets state as int representation of state."""
+    self._validate(int, value, 'state')
+
+  @property
+  def last_modified(self):
```
Comment (sebastienvas) [inline release.py]: datetime ?

File: release.py [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+  def labels(self):
+    return self._release['labels']
+
+  @labels.setter
+  def labels(self, value):
+    """Sets labels as list of string labels."""
+    self._validate_array(basestring, value, 'labels')
+
+  @property
+  def state(self):
```
Comment (sebastienvas) [inline release.py:104]: maybe validate the state ?

File: task.py [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+  def started(self):
+    return self._task['started']
+
+  @started.setter
+  def started(self, value):
+    """Sets started as an int unix datetime."""
+    self._validate(int, value, 'started')
+
+  @property
+  def status(self):
```
Comment (sebastienvas) [inline task.py]: check status

File: task.py [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+
+  def _validate(self, check, value, attribute):
+    """Helper function to validate setter values and set attributes."""
+    if isinstance(value, check):
+      self._task[attribute] = value
+    else:
+      error = 'Invalid input for ' + attribute + ': not a ' + check.__name__
+      raise ValueError(error)
+
+  @property
```
Comment (sebastienvas) [inline task.py]: you might want to cast is ie return str(...)

File: task.py [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+  def task_name(self):
+    return self._task['task_name']
+
+  @task_name.setter
+  def task_name(self, value):
+    """Sets name as string value."""
+    self._validate(basestring, value, 'task_name')
+
+  @property
+  def dependent_on(self):
```
Comment (sebastienvas) [inline task.py]: cast, ie return int()

File: release.py [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+  def started(self):
+    return self._release['started']
+
+  @started.setter
+  def started(self, value):
+    """Sets started as an int unix datetime."""
+    self._validate(int, value, 'started')
+
+  @property
+  def labels(self):
```
Comment (sebastienvas) [inline release.py]: please cast your properties, ie return str(...)

File: release.py [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+class Release(object):
+  """Class for Release Object."""
+
+  def __init__(self, data=None):
+    """Initializes a release.
+
+    Args:
+      data: dictionary with all the release data (optional).
+    """
+    self._release = data or {}
```
Comment (sebastienvas) [inline release.py]: so setters checks the data but not the constructor so you might want to cast all the properties.

Comment (sebastienvas) [reply release.py]: and you might want to validate as well from the constructor

Comment (cabreraem) [reply release.py]: if I validate all the properties in the constructor and the setters, should I still cast in the get methods or assume all the types will be correct?

**Review by sebastienvas — APPROVED** [review]

**Comment by sebastienvas** [comment]
Thanks I am merging this

**Pull request merged**